### PR TITLE
[READY]Gives the clown a pie cannon, also improves pneumatic cannons

### PR DESCRIPTION
**Pull request by silicons**
Honk!
Kor: This increases FUN!!
Pneumatic cannons improved. They can now be edited to fire sequentially instead of all at once, and can throw beyond your screen range. If you set its range multiplier to more than 1, it'll do that, and the pathing will be similar to non pixel projectiles.
image

### CODE REVIEW

**Comment by The-clown-known-as-Pepe** [comment]
# _**honk**_

**Comment by GunHog** [comment]
I am cool with this, as long as it does not stun people!

**Comment by kevinz000** [comment]
@GunHog only if you use a banana cream pie, lol.

**Comment by RemieRichards** [comment]
This was foretold in the legends.

File: code/game/objects/items/weapons/clown.dm [hunk 1]
```diff
@@ -0,0 +1,44 @@
+
+/obj/item/weapon/pie_cannon
```
Comment (RemieRichards) [inline code/game/objects/items/weapons/clown.dm]: this is kind of disgusting tbh.

It should really just be a special pneumatic cannon that only takes pies (then you don't need this projectile nonsense)

**Comment by AnturK** [comment]
Can only load one pie 👎

**Comment by RemieRichards** [comment]
Fixing it so it subtypes from pneumatic cannons would allow it to hold multiple pies :^)

**Comment by kevinz000** [comment]
@RemieRichards I kinda did it on purpose because projectiles are better in terms of aiming and otherwise, as pneumatic cannons are throwcode and that's way less accurate than projectiles...

**Comment by RemieRichards** [comment]
@kevinz000 make it a pneumatic cannon or it's a tablin' for you.

**Comment by kevinz000** [comment]
:(

**Comment by kevinz000** [comment]
wip while i make the targetting not shit

**Comment by FantasticFwoosh** [comment]
Im not sure the clown needs more incentive than what they already have in dragging the cream pie locker/bag purely full of pies around with them fully stocked & stun locking people by throwing as it is.

**Comment by kevinz000** [comment]
@FantasticFwoosh fun hater

**Comment by AnturK** [comment]
How about you fix pneumatic cannons instead of making snowflake type

**Comment by RemieRichards** [comment]
Uhhhh, @AnturK

**Comment by AnturK** [comment]
I'm too slow with refreshing pages.

**Comment by kevinz000** [comment]
lul

**Comment by Bawhoppen** [comment]
wasn't this a joke in the ideas subforum description at one point?

**Comment by RemieRichards** [comment]
@Bawhoppen 
>This was foretold in the legends.

**Comment by Bawhoppen** [comment]
I saw your comment but I couldn't be sure if that we were on the same page

**Comment by RemieRichards** [comment]
fair enough, I was referring to that yes.

**Comment by FantasticFwoosh** [comment]
This would be as powerful as a HONK mech weapon, why hasn't this PR been closed down already?

**Comment by kevinz000** [comment]
@Bawhoppen i did this because i witnessed a bagil round where the clown got sued for cream pieing someone
now imagine if he had a roundstart cannon to do it with?

File: code/game/objects/items/weapons/pneumaticCannon.dm [hunk 2]
```diff
@@ -109,17 +113,53 @@
...
 		user.Weaken(3)
 
+/obj/item/weapon/pneumatic_cannon/proc/fire_items(turf/target, mob/user)
+	switch(fire_mode)
+		if(PCANNON_FIREALL)
+			for(var/obj/item/ITD in loadedItems) //Item To Discharge
+				if(!throw_item(target, ITD, user))
+					break
+		if(PCANNON_FILO)
+			for(var/i = 0, i < throw_amount, i++)
```
Comment (RemieRichards) [inline code/game/objects/items/weapons/pneumaticCannon.dm]: You know what's wrong here.

Comment (kevinz000) [inline code/game/objects/items/weapons/pneumaticCannon.dm]: is it the two modes having nearly the exact same code and still being separated?

File: code/game/objects/items/weapons/pneumaticCannon.dm [hunk 3]
```diff
@@ -134,25 +134,19 @@
...
-					break
-				var/obj/item/I = loadedItems[1]
-				if(!throw_item(target, I, user))
-					break
+	if(fire_mode == PCANNON_FIREALL)
+		for(var/obj/item/ITD in loadedItems) //Item To Discharge
+			if(!throw_item(target, ITD, user))
+				break
+	else
+		for(var/i = 0, i < throw_amount, i++)
```
Comment (RemieRichards) [inline code/game/objects/items/weapons/pneumaticCannon.dm]: and this

Comment (kevinz000) [inline code/game/objects/items/weapons/pneumaticCannon.dm]: ........ uhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhh im assuming its a bad thing that i can't tell what's wrong..

Comment (RemieRichards) [inline code/game/objects/items/weapons/pneumaticCannon.dm]: in to, loop please.

**Comment by kevinz000** [comment]
whoop it has runtimes =v

File: code/game/objects/items/weapons/pneumaticCannon.dm [hunk 4]
```diff
@@ -164,3 +210,27 @@
 		return
 	add_overlay(tank.icon_state)
 	src.update_icon()
+
+/obj/item/weapon/pneumatic_cannon/proc/fill_with_type(type, amount)
+	if(!ispath(type, /obj/item))
+		return FALSE
+	for(var/i = 0, i < amount, i++)
```
Comment (RemieRichards) [inline code/game/objects/items/weapons/pneumaticCannon.dm]: *cough*

Comment (kevinz000) [inline code/game/objects/items/weapons/pneumaticCannon.dm]: wooops

**Comment by RemieRichards** [comment]
Due to the clumsy check, the clown will hit themselves with the pies 75% of the time.

**Comment by kevinz000** [comment]
oh woops let me fix that

**Comment by RemieRichards** [comment]
Also, one thing about ```X in Y to Z``` loops is that they're INCLUSIVE, meaning they match ```<=```'s not ```<```'s like you had before, just give them a once over to make sure they're still doing what you wanted.

**Comment by kevinz000** [comment]
Alright, thanks.

**Comment by Tacolizard** [comment]
the sprite sux

**Comment by kevinz000** [comment]
@Tacolizard feel free to give me new ones.

**Comment by Haevacht** [comment]
that sprite is not garish enough for a clown weapon

File: code/game/objects/items/weapons/pneumaticCannon.dm [hunk 5]
```diff
@@ -16,7 +20,15 @@
 	var/gasPerThrow = 3 //How much gas is drawn from a tank's pressure to fire
 	var/list/loadedItems = list() //The items loaded into the cannon that will be fired out
 	var/pressureSetting = 1 //How powerful the cannon is - higher pressure = more gas but more powerful throws
+	var/checktank = TRUE
+	var/range_multiplier = 1
+	var/throw_amount = 20	//How many items to throw per fire
+	var/fire_mode = PCANNON_FIREALL
+	var/automatic = FALSE
+	var/clumsyCheck = TRUE
```
Comment (Cyberboss) [inline code/game/objects/items/weapons/pneumaticCannon.dm]: Badmin verbs detected

File: code/game/objects/items/weapons/pneumaticCannon.dm [hunk 6]
```diff
@@ -55,19 +67,26 @@
...
-		loadedWeightClass += IW.w_class
-
+		load_item(IW, user)
 
+/obj/item/weapon/pneumatic_cannon/proc/load_item(obj/item/I, mob/user)
+	if((loadedWeightClass + I.w_class) > maxWeightClass)	//Only make messages if there's a user
+		if(user)
+			to_chat(user, "<span class='warning'>\The [I] won't fit into \the [src]!</span>")
+		return FALSE
+	if(I.w_class > src.w_class)
```
Comment (Cyberboss) [inline code/game/objects/items/weapons/pneumaticCannon.dm]: remove src

File: code/game/objects/items/weapons/pneumaticCannon.dm [hunk 7]
```diff
@@ -55,19 +67,26 @@
...
+		if(user)
+			to_chat(user, "<span class='warning'>\The [I] is too large to fit into \the [src]!</span>")
+			return FALSE
+	if(user)		//Only use transfer proc if there's a user, otherwise just set loc.
+		if(!user.transferItemToLoc(I, src))
+			return FALSE
+		to_chat(user, "<span class='notice'>You load \the [I] into \the [src].</span>")
+	else
+		I.forceMove(src)
+	loadedItems.Add(I)
```
Comment (Cyberboss) [inline code/game/objects/items/weapons/pneumaticCannon.dm]: +=

File: code/game/objects/items/weapons/pneumaticCannon.dm [hunk 8]
```diff
@@ -109,17 +127,46 @@
...
+				I = loadedItems[loadedItems.len]
+			if(!throw_item(target, I, user))
+				break
+
+/obj/item/weapon/pneumatic_cannon/proc/throw_item(turf/target, obj/item/I, mob/user)
+	if(!istype(I))
+		return FALSE
+	loadedItems.Remove(I)
+	loadedWeightClass -= I.w_class
+	I.loc = get_turf(src)
```
Comment (Cyberboss) [inline code/game/objects/items/weapons/pneumaticCannon.dm]: forcemove

File: code/game/objects/items/weapons/pneumaticCannon.dm [hunk 9]
```diff
@@ -109,17 +127,46 @@
...
+			var/obj/item/I = loadedItems[1]
+			if(fire_mode == PCANNON_FILO)
+				I = loadedItems[loadedItems.len]
+			if(!throw_item(target, I, user))
+				break
+
+/obj/item/weapon/pneumatic_cannon/proc/throw_item(turf/target, obj/item/I, mob/user)
+	if(!istype(I))
+		return FALSE
+	loadedItems.Remove(I)
```
Comment (Cyberboss) [inline code/game/objects/items/weapons/pneumaticCannon.dm]: -=

File: code/game/objects/items/weapons/pneumaticCannon.dm [hunk 10]
```diff
@@ -164,3 +211,28 @@
...
 	add_overlay(tank.icon_state)
 	src.update_icon()
+
+/obj/item/weapon/pneumatic_cannon/proc/fill_with_type(type, amount)
+	if(!ispath(type, /obj/item))
+		return FALSE
+	for(var/i in 1 to amount)
+		var/obj/item/I = new type
+		if(!load_item(I, null))
+			return TRUE
```
Comment (Cyberboss) [inline code/game/objects/items/weapons/pneumaticCannon.dm]: You need to qdel I if this check fails

File: code/game/objects/items/weapons/pneumaticCannon.dm [hunk 11]
```diff
@@ -109,17 +127,46 @@
...
+/obj/item/weapon/pneumatic_cannon/proc/fire_items(turf/target, mob/user)
+	if(fire_mode == PCANNON_FIREALL)
+		for(var/obj/item/ITD in loadedItems) //Item To Discharge
+			if(!throw_item(target, ITD, user))
+				break
+	else
+		for(var/i in 1 to throw_amount)
+			if(!loadedItems.len)
+				break
+			var/obj/item/I = loadedItems[1]
```
Comment (Cyberboss) [inline code/game/objects/items/weapons/pneumaticCannon.dm]: Put this assignment in the else of the if on the next line

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/game/objects/items/weapons/pneumaticCannon.dm [hunk 12]
```diff
@@ -164,3 +211,28 @@
...
+	checktank = FALSE
+	range_multiplier = 3
+	fire_mode = PCANNON_FIFO
+	throw_amount = 1
+	maxWeightClass = 100	//50 pies. :^)
+	clumsyCheck = FALSE
+
+/obj/item/weapon/pneumatic_cannon/pie/attackby(obj/item/I, mob/living/L)
+	if(istype(I, /obj/item/weapon/reagent_containers/food/snacks/pie))
+		return ..()
```
Comment (Cyberboss) [inline code/game/objects/items/weapons/pneumaticCannon.dm]: Two refactors are needed add a `CanLoadItemProc` and remove this attackby.
ALSO the parent's attackby doesn't check for harm intent

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Pull request merged**